src/main/java/org/jenkinsci/plugins/workflow/graphanalysis/ForkScanner.java

```diff
@@ -376,13 +352,17 @@ ArrayDeque<ParallelBlockStart> leastCommonAncestor(@NonNull final Set<FlowNode>
         };
 
         for (FlowNode f : heads) {
-            iterators.add(new FilteratorImpl<>((Iterator)f.iterateEnclosingBlocks().iterator(), notAHead));  // We can do this because Parallels always meet at a BlockStartNode
+            iterators.add(new FilteratorImpl<>((Iterator) new ExpandedEnclosingBlockIterator(f), notAHead));  // We can do this because Parallels always meet at a BlockStartNode
             FlowSegment b = new FlowSegment();
             b.add(f);
             livePieces.add(b);
             branches.put(f, b);
         }
 
+        // We want to visit heads in order of most recent to least recent.
+        Collections.reverse(iterators);
+        Collections.reverse(livePieces);
```

Comment:
This is the first issue with the ordering, we were looking at heads in the wrong order.